added fixtures as argument to all the test cases

nissampa opened the pull request:
Added new topology - smartswitch-t1 
Added same as pytest marker to directory smartswitch.


-------------------------------------------- live log sessionfinish ---------------------------------------------
00:56:35 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
============================================ short test summary info ============================================
SKIPPED [1] smartswitch/test_show_platform_dpu.py:21: Test is not supported for this testbed and os version
======================================== 1 skipped, 1 warning in 55.71s =========================================
root@sonic-ucs-m6-17:/data/tests# 


*********************************************************************************************************************


-------------------------------------------- live log sessionfinish ---------------------------------------------
01:02:46 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
============================================ short test summary info ============================================
SKIPPED [1] smartswitch/test_show_platform_dpu.py: test requires topology in Mark(name='topology', args=('smartswitch',), kwargs={})

vvolam commented:
@nissampa, Can you add a topology for "smartswitch" similar to below:
-----------------------------
# This test only runs on t2 systems.
pytestmark = [
    pytest.mark.topology('t2')
]
-----------------------------

so that similar check can be added for all future files instead of adding parameter to each test?

nissampa commented:
> @nissampa, Can you add a topology for "smartswitch" similar to below:
> 
> # This test only runs on t2 systems.
> pytestmark = [ pytest.mark.topology('t2') ]
> 
> so that similar check can be added for all future files instead of adding parameter to each test?

- Added smartswitch topo by copying t1-28 lag topo.
- Changed pytest marker as well to use smartswitch.

oleksandrivantsiv commented:
@nissampa Is the SmartSwitch topology for T0, T1, or both? Why do we need to have special handling for a SmartSwitch rather than just checking the platform.json file to identify if the device is a SmartSwitch?

prgeor commented:
> @nissampa Is the SmartSwitch topology for T0, T1, or both? Why do we need to have special handling for a SmartSwitch rather than just checking the platform.json file to identify if the device is a SmartSwitch?

@oleksandrivantsiv we want to keep the source of truth outside the sonic image version. @nissampa I think we need to define new field `is_smartswitch` under `ansible/testbed.yaml` which if True, will indicate the test to run on smartswitch. Absence or False of this field mean not smartswitch. Agree with @oleksandrivantsiv  that this new topology will prevent running smartswitch platform test on T1 topologies.

oleksandrivantsiv commented:
@congh-nvidia please review

congh-nvidia commented:
@nissampa I think it's better to check config_db DEVICE_METADATA "subtype" == "SmartSwitch" to tell whether we have the smartswitch configuration and topology applied. 
And I agree with @prgeor that we can have the is_smartswitch under ansible/testbed.yaml, if it's true, we apply the smartswitch basic configurations(such as midplane interface, DPU, etc), although currently we don't have a method to apply those via minigraph.

JibinBao reviewed (commented):
Since the fixture of check_smartswitch_and_dark_mode will be applied to all tests in the folder of smartswitch. suggest change the scop of check_smartswitch_and_dark_mode from function to session, and add autouse=True. So, we don't need to update the case one by one.

@pytest.fixture(scope='session',  autouse=True)
def check_smartswitch_and_dark_mode

nissampa commented:
> scope='session', autouse=True

Changed the scope.

nissampa commented:
> `is_smartswitch`

Agreed and added the change in ansible/testbed.yaml to have is_smartswitch filed.

nissampa commented:
> @nissampa I think it's better to check config_db DEVICE_METADATA "subtype" == "SmartSwitch" to tell whether we have the smartswitch configuration and topology applied. And I agree with @prgeor that we can have the is_smartswitch under ansible/testbed.yaml, if it's true, we apply the smartswitch basic configurations(such as midplane interface, DPU, etc), although currently we don't have a method to apply those via minigraph.

I agreed with second part and added is_smartswitch.
Regarding the first part, we have discussed and added check_smartswitch fixtures for doing the same to check smartswitch.

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -17,12 +17,13 @@
 ]
 
 
 def test_dpu_ping_after_reboot(duthosts, enum_rand_one_per_hwsku_hostname,
```
enum_rand_one_per_hwsku_hostname is of scope "module" and check_smartswitch_and_dark_mode is of scope "session". You might have to modify either of it to fix scope mismatch.

nissampa replied:
Changed check_smartswitch_and_dark_mode to be scope=module. Sorry for the confusion. I just copy pasted jinbao's suggestion. Realized now.

JibinBao commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -12,12 +12,12 @@
 
 def test_midplane_ip(duthosts, enum_rand_one_per_hwsku_hostname,
-                     platform_api_conn):
+                     platform_api_conn, check_smartswitch_and_dark_mode):
```
Since the autouse of  check_smartswitch_and_dark_mode is True, we don't need to call it explicitly here.
I think all test can remove the argument of check_smartswitch_and_dark_mode

nissampa replied:
Correct. I will remove it.

JibinBao commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -23,7 +23,7 @@ def num_dpu_modules(platform_api_conn):
 
 
-@pytest.fixture(scope='function')
+@pytest.fixture(scope='function', autouse=True)
```
Can we change the scope to module?  so, it only run one time for the tests in one module

nissampa replied:
Making it as module scope resulted in error as follows.
It tries to use platform_api_conn (another fixture) which is of functional scope.

Example:
ERROR smartswitch/platform_tests/test_show_platform_dpu.py::test_midplane_ip[str2-7050cx3-acs-03] - Failed: ScopeMismatch: You tried to access the function scoped fixture platform_api_conn with a module scoped request object, involved factories:

JibinBao replied:
Got it. Thanks

nissampa commented on ansible/testbed.yaml:
```diff
@@ -13,6 +13,7 @@
     - str-msn2700-01
   inv_name: lab
   auto_recover: 'False'
+  is_smartswitch: 'False'
```
@prgeor, Can we have this removed as cough suggested ?

congh-nvidia commented on ansible/vars/topo_smartswitch.yml:
```diff
@@ -0,0 +1,669 @@
+topology:
```
@nissampa could you please rename the file to be like topo_smartswitch-t1.yml? So it will be easier to name the t0 smartswitch topo in future. And we also need the symbolic links for the eos VM tor and leaf template, like ansible/roles/eos/templates/t1-28-lag-tor.j2 and ansible/roles/eos/templates/t1-28-lag-leaf.j2.

nissampa replied:
Changed the topo name to topo_smartswitch-t1.
Added symbolic links as well. 

Please let me know if there is anywhere I need to add topo in any file.

vvolam commented:
@nissampa  could you update your verification information in description?

Also, could you make sure t1 topology tests on smartswitch?

nissampa commented:
> @nissampa could you update your verification information in description?
> 
> Also, could you make sure t1 topology tests on smartswitch?

Updated description.
Checked t1 topology runs on smartswitch.
Checked tests under smartswitch directory are skipping for t1 topo tests.

nissampa commented:
Kindly let me know if it's good to approve and merge.

nissampa commented:
@congh-nvidia, @prgeor @JibinBao 

If all looks good, can you please approve and merge it ?

vvolam commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -85,10 +85,9 @@ def dpu_power_on(duthost, platform_api_conn):
         Returns True or False based on all DPUs powered on or not
     """
 
-    num_modules = num_dpu_modules(platform_api_conn)
```
do we need to add num_dpu_modules to the function?

nissampa replied:
Yes, it was given as comment in last PR.

vvolam commented:
> @congh-nvidia, @prgeor @JibinBao
> 
> If all looks good, can you please approve and merge it ?

@nissampa getting these errors when running on non-smartswitch

23:31:33 __init__._fixture_func_decorator         L0073 ERROR  |
KeyError('DPUS')
Traceback (most recent call last):
  File "/var/src/src/sonic-mgmt-int/tests/common/plugins/log_section_start/__init__.py", line 71, in _fixture_func_decorator
    return fixture_func(*args, **kargs)
  File "/var/src/src/sonic-mgmt-int/tests/smartswitch/common/device_utils_dpu.py", line 41, in check_smartswitch_and_dark_mode
    if not duthost.facts["DPUS"] and \
KeyError: 'DPUS'

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...............................................................Passed<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/smartswitch/common/device_utils_dpu.py:40:24: F405 're' may be undefined, or defined from star imports: tests.common.devices.sonic, tests.platform_tests.api.conftest<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..............................................Passed<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...............................................................Passed<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/smartswitch/common/device_utils_dpu.py:13:1: E302 expected 2 blank lines, found 1<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..............................................Passed<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

nissampa commented:
> > @congh-nvidia, @prgeor @JibinBao
> > If all looks good, can you please approve and merge it ?
> 
> @nissampa getting these errors when running on non-smartswitch
> 
> 23:31:33 **init**._fixture_func_decorator L0073 ERROR | KeyError('DPUS') Traceback (most recent call last): File "/var/src/src/sonic-mgmt-int/tests/common/plugins/log_section_start/**init**.py", line 71, in _fixture_func_decorator return fixture_func(*args, **kargs) File "/var/src/src/sonic-mgmt-int/tests/smartswitch/common/device_utils_dpu.py", line 41, in check_smartswitch_and_dark_mode if not duthost.facts["DPUS"] and KeyError: 'DPUS'

Resolved it.

congh-nvidia commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -37,29 +37,31 @@ def check_smartswitch_and_dark_mode(duthosts,
 
     duthost = duthosts[enum_rand_one_per_hwsku_hostname]
+    output = duthost.os_version
+    os_version = float(re.search(r'([1-9]{1}\d*)(\.\d{0,2})', output).group())
```
this pattern is not working with our internal images, and it is also not working with master branch. Could  you use the existing skip_release method? https://github.com/sonic-net/sonic-mgmt/blob/9c124855d7d70a121deda59978e710cd06cf2ff2/tests/common/utilities.py#L72

nissampa replied:
Actually, smartswitch case is little different, it can be run on os versions greater than 202405.
The check of DPUS field in platform.json is enough as well to check that.

nissampa replied:
So removed this check altogether

vvolam commented on ansible/testbed.yaml:
```diff
@@ -104,6 +104,7 @@
     - lab-a7260-01
   inv_name: lab
   auto_recover: 'True'
+  is_smartswitch: 'False'
```
Can we make is_smartswitch to False by default and set true only for smart switch testbeds.

nissampa replied:
This one I forgot to remove. 
False and field not present are the same. The check will take care of it according to the previous comment given by prince/congh

vvolam commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -8,7 +8,7 @@
 from tests.common.utilities import wait_until
 from tests.common.helpers.assertions import pytest_assert
-from pkg_resources import parse_version
+import re
```
can you move this import together with all imports in the beginning.

nissampa replied:
removed it as it is not needed now.

vvolam commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -37,29 +37,31 @@ def check_smartswitch_and_dark_mode(duthosts,
 
     duthost = duthosts[enum_rand_one_per_hwsku_hostname]
+    output = duthost.os_version
+    os_version = float(re.search(r'([1-9]{1}\d*)(\.\d{0,2})', output).group())
```
you can simply do this way "os_version = duthost.os_version.split(".")[0]" and compare directly like os_version <= "202405"

congh-nvidia replied:
this will not work with master branch, it's better to use skip_release same as in other tests.

nissampa replied:
Actually, smartswitch case is little different, it can be run on os versions greater than 202405.
The check of DPUS field in platform.json is enough as well to check that. Since DPUS filed will be present in those releases.

congh-nvidia commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -17,18 +17,18 @@
 
 pytestmark = [
-    pytest.mark.topology('t1')
+    pytest.mark.topology('smartswitch-t1')
```
the marker should be 'smartswitch', like pytest.mark.topology('t1') for all t1 topos and pytest.mark.topology('t0') for all t0.

nissampa replied:
Changed it.

vvolam commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -37,29 +37,31 @@ def check_smartswitch_and_dark_mode(duthosts,
+                                    num_dpu_modules)
 
     if darkmode:
         dpu_power_on(duthost, platform_api_conn)
```
dpu_power_on also takes num_dpu_modules as argument.

nissampa replied:
Changed it.

congh-nvidia commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -12,7 +12,7 @@
 
 pytestmark = [
-    pytest.mark.topology('t1')
+    pytest.mark.topology('smartswitch-t1')
```
the marker should be 'smartswitch', like pytest.mark.topology('t1') for all t1 topos and pytest.mark.topology('t0') for all t0.

nissampa replied:
Changed it. Since you mentioned to change the yaml file to smartswitch-t1, I used marker as well to be changed. now changed back the marker to just smartswitch.

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...............................................................Passed<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/smartswitch/common/device_utils_dpu.py:11:1: F401 're' imported but unused<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..............................................Passed<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

vvolam commented:
@nissampa could you please resolve build errors?

nissampa commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -38,28 +37,27 @@ def check_smartswitch_and_dark_mode(duthosts,
-    if not duthost.facts["DPUS"] and \
-            parse_version(duthost.os_version) <= parse_version("202405"):
+    if "DPUS" not in duthost.facts:
         pytest.skip("Test is not supported for this testbed and os version")
```
Yes changed it.

nissampa commented:
/AzurePipelines run

nissampa commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

bpar9 reviewed (approved):
lgtm
